certbot/tests/client_test.py

```diff
@@ -25,15 +27,29 @@ def setUp(self):
         self.config.register_unsafely_without_email = False
         self.config.email = "alias@example.com"
         self.account_storage = account.AccountMemoryStorage()
+        self.config.eab_hmac_key = "hmac-key-for-testing"
+        self.config.eab_kid = "kid-for-testing"
 
     def _call(self):
         from certbot.client import register
         tos_cb = mock.MagicMock()
         return register(self.config, self.account_storage, tos_cb)
 
+    @staticmethod
+    def _public_key_mock():
+        m = mock.Mock(__class__=interfaces.JSONDeSerializable)
+        m.to_partial_json.return_value = '{"a": 1}'
+        return m
+
+    @staticmethod
+    def _directory_getitem_mock(input):
+        return "/acme/new-account"
+
     def test_no_tos(self):
         with mock.patch("certbot.client.acme_client.BackwardsCompatibleClientV2") as mock_client:
             mock_client.new_account_and_tos().terms_of_service = "http://tos"
+            mock_client().client.net.key.public_key = mock.Mock(side_effect=self._public_key_mock)
```

Comment:
Why do we now need these? Can we find a way to test, such that we don't?

Reply:
The reason I put it there, after some discussions with @joohoi and @adferrand on #letsencrypt-dev, is because of:

https://github.com/robaman/certbot/blob/b3e75257a7bc2ea53983efb60ee9baae1ca74bc4/certbot/client.py#L209

Without it, it returns a mock that does not support to_partial_json() and need to be an interfaces.JSONDeSerializable because of a isinstance() check.

The mock below it is because of:
https://github.com/robaman/certbot/blob/b3e75257a7bc2ea53983efb60ee9baae1ca74bc4/acme/acme/messages.py#L274

...not returning an URL correctly.



I've been thinking about other approaches to do it, but I can't find out another way.

Do you have any pointers for me?

Reply:
The easiest way is to not have `config.eab_kid and config.eab_hmac_key` exist, by removing them from `setUp` and making sure that if `config` is mocked, they return `None`, again in `setUp`. And then setting them to have values explicitly in functions that are testing EAB.

Reply:
Hmm. Was this answer supposed to be on another thread? Where you started saying "We want these to sometimes not be available, so they should be set in config only when explicitly testing that an EAB is constructed." perhaps?

I still haven't figured out how to proceed with the mock stuff.

Reply:
It's relevant to both -- if you do that, it should also fix this problem, because it won't hit the EAB stuff in the first place.

Reply:
Think I understood now. :)